File: eva-core/src/main/java/io/nerv/core/mvc/util/Page.java
package io.nerv.core.mvc.util;

import com.baomidou.mybatisplus.core.metadata.IPage;
import com.baomidou.mybatisplus.core.metadata.OrderItem;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.function.Predicate;

/**
 * 自定义分页模型
 * @author PKAQ
 */
public class Page<T> implements IPage<T> {

    private static final long serialVersionUID = 8545996863226528798L;

    /**
     * 查询数据列表
     */
    private List<T> list = Collections.emptyList();

    /**
     * 总数
     */
    private long total = 0;
    /**
     * 每页显示条数，默认 10
     */
    private long size = 10;

    /**
     * 当前页
     */
    private long current = 1;

    /**
     * 排序字段信息
     */
    private List<OrderItem> orders = new ArrayList<>();

    /**
     * 自动优化 COUNT SQL
     */
    private boolean optimizeCountSql = true;
    /**
     * 是否进行 count 查询
     */
    private boolean isSearchCount = true;
    /**
     * 是否命中count缓存
     */
    private boolean hitCount = false;

    public Page() {

    }

    /**
     * 分页构造函数
     *
     * @param current 当前页
     * @param size    每页显示条数
     */
    public Page(long current, long size) {
        this(current, size, 0);
    }

    public Page(long current, long size, long total) {
        this(current, size, total, true);
    }

    public Page(long current, long size, boolean isSearchCount) {
        this(current, size, 0, isSearchCount);
    }

    public Page(long current, long size, long total, boolean isSearchCount) {
        if (current > 1) {
            this.current = current;
        }
        this.size = size;
        this.total = total;
        this.isSearchCount = isSearchCount;
    }

    /**
     * 是否存在上一页
     *
     * @return true / false
     */
    public boolean hasPrevious() {
        return this.current > 1;
    }

    /**
     * 是否存在下一页
     *
     * @return true / false
     */
    public boolean hasNext() {
        return this.current < this.getPages();
    }

    @Override
    public List<T> getRecords() {
        return null;
    }

    public List<T> getList() {
        return this.list;
    }

    @Override
    public Page<T> setRecords(List<T> data) {
        this.list = data;
        return this;
    }

    @Override
    public long getTotal() {
        return this.total;
    }

    @Override
    public Page<T> setTotal(long total) {
        this.total = total;
        return this;
    }

    @Override
    public long getSize() {
        return this.size;
    }

    @Override
    public Page<T> setSize(long size) {
        this.size = size;
        return this;
    }

    @Override
    public long getCurrent() {
        return this.current;
    }

    @Override
    public Page<T> setCurrent(long current) {
        this.current = current;
        return this;
    }

    /**
     * 查找 order 中正序排序的字段数组
     *
     * @param filter 过滤器
     * @return 返回正序排列的字段数组
     */
    private String[] mapOrderToArray(Predicate<OrderItem> filter) {
        List<String> columns = new ArrayList<>(orders.size());
        orders.forEach(i -> {
            if (filter.test(i)) {
                columns.add(i.getColumn());
            }
        });
        return columns.toArray(new String[0]);
    }

    /**
     * 移除符合条件的条件
     *
     * @param filter 条件判断
     */
    private void removeOrder(Predicate<OrderItem> filter) {
        for (int i = orders.size() - 1; i >= 0; i--) {
            if (filter.test(orders.get(i))) {
                orders.remove(i);
            }
        }
    }

    /**
     * 添加新的排序条件，构造条件可以使用工厂：{@link OrderItem#(String, boolean)}
     *
     * @param items 条件
     * @return 返回分页参数本身
     */
    public Page<T> addOrder(OrderItem... items) {
        orders.addAll(Arrays.asList(items));
        return this;
    }

    /**
     * 添加新的排序条件，构造条件可以使用工厂：{@link OrderItem#(String, boolean)}
     *
     * @param items 条件
     * @return 返回分页参数本身
     */
    public Page<T> addOrder(List<OrderItem> items) {
        orders.addAll(items);
        return this;
    }

    @Override
    public List<OrderItem> orders() {
        return getOrders();
    }

    public List<OrderItem> getOrders() {
        return orders;
    }

    public void setOrders(List<OrderItem> orders) {
        this.orders = orders;
    }

    @Override
    public boolean optimizeCountSql() {
        return optimizeCountSql;
    }

    @Override
    public boolean isSearchCount() {
        if (total < 0) {
            return false;
        }
        return isSearchCount;
    }

    public Page<T> setSearchCount(boolean isSearchCount) {
        this.isSearchCount = isSearchCount;
        return this;
    }

    public Page<T> setOptimizeCountSql(boolean optimizeCountSql) {
        this.optimizeCountSql = optimizeCountSql;
        return this;
    }

    @Override
    public void hitCount(boolean hit) {
        this.hitCount = hit;
    }

    @Override
    public boolean isHitCount() {
        return hitCount;
    }
}

File: eva-core/src/main/kotlin/io/nerv/core/mvc/util/Page.kt
package io.nerv.core.mvc.util

import com.baomidou.mybatisplus.core.metadata.IPage
import com.baomidou.mybatisplus.core.metadata.OrderItem
import java.util.*
import java.util.function.Consumer
import java.util.function.Predicate

/**
 * 自定义分页模型
 * @author PKAQ
 */
open class Page<T> : IPage<T> {
    /**
     * 查询数据列表
     */
    var list: List<T> = emptyList()
        set

    /**
     * 总数
     */
    var total: Number = 0

    /**
     * 每页显示条数，默认 10
     */
    var size: Number = 10

    /**
     * 当前页
     */
    var current: Number = 1

    /**
     * 排序字段信息
     */
    var orders: MutableList<OrderItem> = ArrayList()

    /**
     * 自动优化 COUNT SQL
     */
    var optimizeCountSql = true

    /**
     * 是否进行 count 查询
     */
    var searchCount: Boolean = true

    /**
     * 是否命中count缓存
     */
    var hitCount = false

    constructor() {}
    constructor(current: Long, size: Long, isSearchCount: Boolean) : this(current, size, 0, isSearchCount) {}

    /**
     * 分页构造函数
     *
     * @param current 当前页
     * @param size    每页显示条数
     */
    @JvmOverloads
    constructor(current: Long, size: Long, total: Long = 0, isSearchCount: Boolean = true) {
        if (current > 1) {
            this.current = current
        }
        this.size = size
        this.total = total
        this.isSearchCount = isSearchCount
    }

    /**
     * 是否存在上一页
     *
     * @return true / false
     */
    fun hasPrevious(): Boolean {
        return current > 1
    }

    /**
     * 是否存在下一页
     *
     * @return true / false
     */
    operator fun hasNext(): Boolean {
        return current < this.pages
    }

    override fun getRecords(): List<T> {
        return null
    }

    override fun setRecords(data: List<T>): Page<T> {
        list = data
        return this
    }

    override fun getTotal(): Long {
        return total
    }

    override fun setTotal(total: Long): Page<T> {
        this.total = total
        return this
    }

    override fun getSize(): Long {
        return size
    }

    override fun setSize(size: Long): Page<T> {
        this.size = size
        return this
    }

    override fun getCurrent(): Long {
        return current
    }

    override fun setCurrent(current: Long): Page<T> {
        this.current = current
        return this
    }

    /**
     * 查找 order 中正序排序的字段数组
     *
     * @param filter 过滤器
     * @return 返回正序排列的字段数组
     */
    fun mapOrderToArray(filter: Predicate<OrderItem>): Array<String> {
        val columns: MutableList<String> = ArrayList(orders.size)
        orders.forEach(Consumer { i: OrderItem ->
            if (filter.test(i)) {
                columns.add(i.column)
            }
        })
        return columns.toTypedArray()
    }

    /**
     * 移除符合条件的条件
     *
     * @param filter 条件判断
     */
    fun removeOrder(filter: Predicate<OrderItem>) {
        for (i in orders.indices.reversed()) {
            if (filter.test(orders[i])) {
                orders.removeAt(i)
            }
        }
    }

    /**
     * 添加新的排序条件，构造条件可以使用工厂：[OrderItem.]
     *
     * @param items 条件
     * @return 返回分页参数本身
     */
    fun addOrder(vararg items: OrderItem?): Page<T> {
        orders.addAll(Arrays.asList(*items))
        return this
    }

    /**
     * 添加新的排序条件，构造条件可以使用工厂：[OrderItem.]
     *
     * @param items 条件
     * @return 返回分页参数本身
     */
    fun addOrder(items: List<OrderItem>?): Page<T> {
        orders.addAll(items!!)
        return this
    }

    override fun orders(): List<OrderItem> {
        return getOrders()
    }

    fun getOrders(): List<OrderItem> {
        return orders
    }

    fun setOrders(orders: MutableList<OrderItem>) {
        this.orders = orders
    }

    override fun optimizeCountSql(): Boolean {
        return optimizeCountSql
    }

    override fun isSearchCount(): Boolean {
        return if (total < 0) {
            false
        } else isSearchCount
    }

    fun setSearchCount(isSearchCount: Boolean): Page<T> {
        this.isSearchCount = isSearchCount
        return this
    }

    fun setOptimizeCountSql(optimizeCountSql: Boolean): Page<T> {
        this.optimizeCountSql = optimizeCountSql
        return this
    }

    override fun hitCount(hit: Boolean) {
        hitCount = hit
    }

    override fun isHitCount(): Boolean {
        return hitCount
    }

    companion object {
        const val serialVersionUID = 8545996863226528798L
    }
}